[UnifiedPDF] [iPadOS] Web process crash under WebKit::populateCaretContext() when trackpad hover over PDF document

aprotyas opened the pull request:
#### 1113729a76ef64168a293440405f039a6b5977bf
<pre>
[UnifiedPDF] [iPadOS] Web process crash under WebKit::populateCaretContext() when trackpad hover over PDF document
<a href="https://bugs.webkit.org/show_bug.cgi?id=283477">https://bugs.webkit.org/show_bug.cgi?id=283477</a>
<a href="https://rdar.apple.com/140191140">rdar://140191140</a>

Reviewed by Tim Horton, Wenson Hsieh, and Richard Robinson.

Currently, whenever the trackpad hovers over a PDF document, the web
process crashes trying to retrieve position information because of an
unchecked std::optional&lt;Cursor&gt; access. The RenderEmbeddedObject
corresponding to the PDF plugin thinks the plugin should handle cursor
updates, hence returning a null cursor value to populateCaretContext().

This patch avoids this very frequent crasher by ensuring that we check
the optional cursor value in the interaction information argument is
non-empty before reasoning about it. We may revisit this codepath when
we implement cursor update support for Unified PDF on iOS, but at least
this patch avoids the crash.

API test: UnifiedPDF.MouseDidMoveOverPDF

We introduce test coverage by exercising a mouse move over a PDF
document, which would cause the WP to crash without the proposed
changes.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::populateCaretContext):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/MouseSupportUIDelegate.h: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/MouseSupportUIDelegate.mm: Added.
(-[MouseSupportUIDelegate _webView:mouseDidMoveOverElement:withFlags:userInfo:]):
(-[MouseSupportUIDelegate setMouseDidMoveOverElementHandler:]):

Separate MouseSupportUIDelegate into its own header, which makes this
sharable across multiple test suites.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:

Unified sources build fix.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:

Unified sources build fix.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(-[MouseSupportUIDelegate _webView:mouseDidMoveOverElement:withFlags:userInfo:]): Deleted.
(-[MouseSupportUIDelegate setMouseDidMoveOverElementHandler:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/286927@main">https://commits.webkit.org/286927@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77ee575dbf4d73dc7a6504ff81ee012df4a7c7d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77554 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/56589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/30470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/28836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79671 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/65739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4886 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/82151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/28836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80621 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/65739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/30470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/65739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/30470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/27161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/65739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/30470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/83545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/4934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3359 "Build is in progress. Recent messages:") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/83545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/5090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/30470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/83545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/30470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12006 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/4881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/4900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/8335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/6659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh reviewed (approved):
Nit - typo in the commit message:

> trying to retrive position information

retrive → retrieve

aprotyas committed:
[UnifiedPDF] [iPadOS] Web process crash under WebKit::populateCaretContext() when trackpad hover over PDF document
https://bugs.webkit.org/show_bug.cgi?id=283477
rdar://140191140

Reviewed by Tim Horton, Wenson Hsieh, and Richard Robinson.

Currently, whenever the trackpad hovers over a PDF document, the web
process crashes trying to retrieve position information because of an
unchecked std::optional<Cursor> access. The RenderEmbeddedObject
corresponding to the PDF plugin thinks the plugin should handle cursor
updates, hence returning a null cursor value to populateCaretContext().

This patch avoids this very frequent crasher by ensuring that we check
the optional cursor value in the interaction information argument is
non-empty before reasoning about it. We may revisit this codepath when
we implement cursor update support for Unified PDF on iOS, but at least
this patch avoids the crash.

API test: UnifiedPDF.MouseDidMoveOverPDF

We introduce test coverage by exercising a mouse move over a PDF
document, which would cause the WP to crash without the proposed
changes.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::populateCaretContext):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/MouseSupportUIDelegate.h: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/MouseSupportUIDelegate.mm: Added.
(-[MouseSupportUIDelegate _webView:mouseDidMoveOverElement:withFlags:userInfo:]):
(-[MouseSupportUIDelegate setMouseDidMoveOverElementHandler:]):

Separate MouseSupportUIDelegate into its own header, which makes this
sharable across multiple test suites.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:

Unified sources build fix.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:

Unified sources build fix.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(-[MouseSupportUIDelegate _webView:mouseDidMoveOverElement:withFlags:userInfo:]): Deleted.
(-[MouseSupportUIDelegate setMouseDidMoveOverElementHandler:]): Deleted.

Canonical link: https://commits.webkit.org/286927@main
